### PR TITLE
added base analytic function and helpers

### PR DESCRIPTION
**Pull request by maxwellchandler**
@sam-bao These are the initial analytic functions. I have a test section with example calls commented out. You can play with them and see the links they make. If you have some urls that you want to test on can you send me them and I will create the mocks?

### CODE REVIEW

File: R/getAnalytics.R [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+#' @param order order
+#' @param columns columns
+#' @param rows rows
+#' @param base_url string - base address of instance (text before api/ in URL)
+#' @param retry retry
+#' @return data frame with the rows of the response
+
+getAnalytics <-  function(..., dimensions = NULL, filters = NULL, start_date = NULL, end_date = NULL,
+                               order = NULL, columns = NULL, rows = NULL,
+                               base_url = getOption("baseurl"), retry=1){
```
Comment (sam-bao) [inline R/getAnalytics.R:18]: @maxwellchandler nice work on the first iteration. I have some thoughts, but I am still flexible on this so please let me know if you have other ideas.

I think the following signature combined with maybe a slightly modified variant of dForm and fForm helpers will work well.

```
(..., dx = NULL, dx_f = NULL, pe = NULL, pe_f = NULL,
ou = NULL, ou_f = NULL, co = NULL, co_f = NULL, 
ao = NULL, ao_f = NULL,
outputIdScheme = "UID",
base_url = getOption("baseurl"), retry = 1)
```
the `...`  can be used for any other analytics endpoint parameters such as `startDate` or `order` and it also receives the miscellaneaous dimensions that we call with a uid (categories, data element group sets, etc).

So an example call may look like
```
age_category_options = c("uid12345678", 
                   "uid87654321") 
sex_category_option = "uid99999999"
data = some_data

getAnalytics(dx = data$dx, 
             ou = data$ou,
             pe_f = "2020Q2",
             dForm(uid11111111 = age_category_options), 
             fForm(uid22222222 = sex_category_option),
             outputIdScheme = "UID")
```

where `dForm(uid11111111 = age_category_options)` would create a string: `"filter=uid11111111:uid12345678;uid87654321"`

and named arguments would also ultimately just call dFrom or fForm so `dx=data$dx` will be translated into `dForm(dx = data$dx)`

For me that is about as concise and readable as we can make a call to getAnalytics.

Comment (sam-bao) [reply R/getAnalytics.R:18]: Just to clarify what that example call might give

The value for some indicators in data$dx, for the facilities listed in data$ou, filtered to include only 2020Q2, dissagregated by 2 age categories and filtered to only include data for one sex.

**Review by sam-bao — CHANGES_REQUESTED** [review]

**Comment by maxwellchandler** [comment]
@sam-bao 
I have implemented everything you wanted. The only difference is `dForm(uid11111111 = age_category_options)`  is now `uid11111111 %.d% age_category_options` because if not `uid11111111` would get treated like a named argument

**Comment by sam-bao** [comment]
@maxwellchandler is this the expected output?

```
> ages = c("age1", "age2")
> "e485zBiR7vG" %.d% ages
[1] "e485zBiR7vG=age1:dimension" "e485zBiR7vG=age2:dimension"
```

**Comment by maxwellchandler** [comment]
@sam-bao sorry thats fixed now

**Comment by sam-bao** [comment]
@maxwellchandler still seems to be an issue with the %.d% and %.f% functions. The dimension items should be separated by semicolons `/api/33/analytics?dimension=dim-id:dim-item;dim-item&dimension=dim-id:dim-item;dim-item` but below I get comma delimiters.

```
> library(datapackcommons)
> library(magrittr)
> library(datimutils)
> disaggs <- dim_item_sets %>% 
+   dplyr::filter(model_sets %in% c("<1-50+",
+                                   "F/M",
+                                   "kp1",
+                                   "tr_neg")) %>%
+   dplyr::select(dim_item_uid, dim_uid) %>%
+   unique()  %>%
+   stats::na.omit() %>%  # there are some items in dim item sets with no source dimension
+   dplyr::group_by(dim_uid) %>% 
+   dplyr::summarise(dim_item_uids = list(dim_item_uid)) %>% 
+   dplyr::ungroup()
> purrr::map2_chr(disaggs$dim_uid, .y = disaggs$dim_item_uids,
+                   ~ (!!.x) %.d% .y)
[1] "dimension=e485zBiR7vG:Z8MTaDxRBP6,BURHq262iEL,egW0hBcZeD2,RV1ZeOr98rI,tIZRQs0FK5P,QOawCj9oLNS,BDPEHrovntA,K9Cw4402aAh,JqZFtdn1sG3,ftAnvKhxRxl,kyKHoWxTzHR,wRYm1EaHwN6,MsbFixtB8mu,bePcXLCq9Ov"
[2] "dimension=FokGv0LCTYj:Gb0GYkqotaO"                                                                                                                                                            
[3] "dimension=jyUTj5YC3OK:ZOYVg7Hosni,hDBPKTjUPDm"                                                                                                                                                
[4] "dimension=VCCs1f22cOR:CKCeNqWcxXe,qY5T9mKjsz5,eiR4kOjT60r,UjruEYjGK4R,pi7naVKPzSM"                                                                                                            
> purrr::map2_chr(disaggs$dim_uid, .y = disaggs$dim_item_uids,
+                   ~ (!!.x) %.f% .y)
[1] "filter=e485zBiR7vG:Z8MTaDxRBP6,BURHq262iEL,egW0hBcZeD2,RV1ZeOr98rI,tIZRQs0FK5P,QOawCj9oLNS,BDPEHrovntA,K9Cw4402aAh,JqZFtdn1sG3,ftAnvKhxRxl,kyKHoWxTzHR,wRYm1EaHwN6,MsbFixtB8mu,bePcXLCq9Ov"
[2] "filter=FokGv0LCTYj:Gb0GYkqotaO"                                                                                                                                                            
[3] "filter=jyUTj5YC3OK:ZOYVg7Hosni,hDBPKTjUPDm"                                                                                                                                                
[4] "filter=VCCs1f22cOR:CKCeNqWcxXe,qY5T9mKjsz5,eiR4kOjT60r,UjruEYjGK4R,pi7naVKPzSM" 
```

Please add unit tests for these functions. They shouldn't even require mocks.

**Comment by sam-bao** [comment]
@maxwellchandler I am getting an error when I make a call that results in an empty record set:

```
> datimutils::getAnalytics(dx = "S28D0BSZqwR",
+                                pe = "2019Oct",
+                                ou = c(datimutils::getOrgUnits("Angola", name), "OU_GROUP-nwQbMeALRjL"),
+                                c("dimension=e485zBiR7vG:tIZRQs0FK5P;QOawCj9oLNS;BDPEHrovntA;K9Cw4402aAh;JqZFtdn1sG3;ftAnvKhxRxl;kyKHoWxTzHR;wRYm1EaHwN6;MsbFixtB8mu;bePcXLCq9Ov", 
+ "dimension=jyUTj5YC3OK:ZOYVg7Hosni"))
Error in names(x) <- value : 
  'names' attribute [6] must be the same length as the vector [0]
```

I think if no actual data are returned from the analytics endpoint (so all we get back is metadata like column names), then getAnalytics can just return `NULL`

**Comment by maxwellchandler** [comment]
@sam-bao Ok fixed

**Comment by sam-bao** [comment]
@maxwellchandler I still get an error when trying to run:

```
datimutils::getAnalytics(dx = "S28D0BSZqwR",
                         pe = "2019Oct",
                         ou = c(datimutils::getOrgUnits("Angola", name), "OU_GROUP-nwQbMeALRjL"),
                         c("dimension=e485zBiR7vG:tIZRQs0FK5P;QOawCj9oLNS;BDPEHrovntA;K9Cw4402aAh;JqZFtdn1sG3;ftAnvKhxRxl;kyKHoWxTzHR;wRYm1EaHwN6;MsbFixtB8mu;bePcXLCq9Ov", 
                           "dimension=jyUTj5YC3OK:ZOYVg7Hosni"))
```

**Comment by sam-bao** [comment]
@maxwellchandler in some cases the list of dimension item values can be `NULL`

image

I don't see this scenario covered in the unit tests, and based on a code review I am not sure we have these scenarios covered.

From the DHIS2 docs

> To query for data broken down by category option combinations instead of data element totals you can include the category dimension in the query string, for instance like this:

```
/api/33/analytics?dimension=dx:fbfJHSPpUQD;cYeuwXTCPkU
  &dimension=co&dimension=pe:201601&dimension=ou:O6uvpzGd5pu;lc3eMKXaEfw
```

Please have a look and add unit tests and/or logic to handle these cases.

I also don't see any unit tests where you specify dimensions or filters other than the standard ones (dx, pe, ...). Please add some units tests for data element group sets and/or organisation unit group sets.

Perhaps rather than accepting NULL for dimension item values, we could support the keyword `all` so

`co_f = "all"`
`uid11111111 %.d% "all"` 

I don't love use of `"all"` because  _all_ doesn't get passed to the API, but I can't really think of a better option. Any other ideas?

**Comment by maxwellchandler** [comment]
@sam-bao I added in the all option because that I thought that was good. I added some more tests too, should be 100% coverage, but messing around with play pivot tables im having a hard time even getting any data for some of the options.

**Pull request merged**